streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -947,127 +665,428 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
+    final def collectAllWith[S](z: S)(f: (S, B) => S): ZSink[R, E, A, A, S] = widen.collectAllWith(z)(f)
 
     /**
      * Accumulates into a list for as long as incoming values verify predicate `p`.
      */
-    final def collectAllWhile(p: A => Boolean): ZSink[R, E, A, A, List[B]] = sink.collectAllWhile(p)
+    final def collectAllWhile(p: A => Boolean): ZSink[R, E, A, A, List[B]] = widen.collectAllWhile(p)
 
     /**
      * Accumulates into a value of type `S` for as long as incoming values verify predicate `p`.
      */
     final def collectAllWhileWith[S](p: A => Boolean)(z: S)(f: (S, B) => S): ZSink[R, E, A, A, S] =
-      sink.collectAllWhileWith(p)(z)(f)
-
-    /**
-     * Accumulates into a value of type `S` for as long as the state satisfies the predicate `p`.
-     */
-    final def collectWhile[S](z: S)(p: S => Boolean)(f: (S, B) => S): ZSink[R, E, A, A, S] =
-      sink.collectWhile(z)(p)(f)
+      widen.collectAllWhileWith(p)(z)(f)
 
     /**
      * A named alias for `?`.
      */
-    final def optional: ZSink[R, E, A, A, Option[B]] = sink.?
+    final def optional: ZSink[R, E, A, A, Option[B]] = widen.?
 
     /**
      * Produces a sink consuming all the elements of type `A` as long as
      * they verify the predicate `pred`.
      */
-    final def takeWhile(pred: A => Boolean): ZSink[R, E, A, A, B] = sink.takeWhile(pred)
+    final def takeWhile(pred: A => Boolean): ZSink[R, E, A, A, B] = widen.takeWhile(pred)
+
+    /**
+     * Creates a sink that produces values until one verifies
+     * the predicate `f`.
+     *
+     * @note The predicate is only verified when the underlying
+     * signals completion, or when the resulting sink is extracted.
+     * Sinks that never signal completion (e.g. [[ZSink.collectAll]])
+     * will not have the predicate applied to intermediate values.
+     */
+    final def untilOutput(f: B => Boolean): ZSink[R, E, A, A, Option[B]] =
+      widen untilOutput f
+  }
+
+  implicit class InvariantOps[R, E, A0, A, B](private val self: ZSink[R, E, A0, A, B]) extends AnyVal {
+
+    /**
+     * Drops all elements entering the sink for as long as the specified predicate
+     * evaluates to `true`.
+     */
+    final def dropWhile(pred: A => Boolean): ZSink[R, E, A0, A, B] =
+      new ZSink[R, E, A0, A, B] {
+        type State = (self.State, Boolean)
+
+        val initial = self.initial.map((_, true))
+
+        def step(state: State, a: A) =
+          if (!state._2) self.step(state._1, a).map((_, false))
+          else {
+            if (pred(a)) UIO.succeed(state)
+            else self.step(state._1, a).map((_, false))
+          }
+
+        def extract(state: State) = self.extract(state._1)
+
+        def cont(state: State) = self.cont(state._1)
+      }
+
+    /**
+     * Filters the inputs fed to this sink.
+     */
+    def filter(f: A => Boolean): ZSink[R, E, A0, A, B] =
+      self match {
```

Comment:
Nitpicking. Not sure how I feel about having `SinkPure` implementations in this file from a discoverability and maintainability standpoint.

Reply:
Well, since `filter` was moved out to an implicit class to improve inference, it can no longer be overriden in `SinkPure.

Reply:
I'm well aware... we'll live with it. :-)